File: DAL/SubTaskRepository.cs
using Entity;
using System;
using System.Collections.Generic;
using System.Data.SqlClient;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DAL
{
    public class SubTaskRepository : BDRepository<SubTask>
    {
        private readonly TaskRepository taskRepository;
        public SubTaskRepository() { taskRepository = new TaskRepository(); }
        public override bool Delete(int id)
        {
            try
            {
                if (id <= 0) return false;
                bd.OpenConection();
                SqlCommand cmd = new SqlCommand("DELETE FROM [dbo].[SubTask] WHERE Id_SubTask= @Id_SubTask", bd.connection);
                cmd.Parameters.AddWithValue("@Id_SubTask", id);
                return cmd.ExecuteNonQuery() > 0;
            }
            catch (SqlException)
            {
                return false;
            }
            catch (Exception)
            {
                return false;
            }
            finally
            {
                bd.CloseConection();
            }
        }
        public override List<SubTask> GetAll()
        {
            List<SubTask> subTasks = new List<SubTask>();
            try
            {
                bd.OpenConection();
                SqlCommand cmd = new SqlCommand("SELECT * FROM [dbo].[SubTask]", bd.connection);
                using (var reader = cmd.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        subTasks.Add(MappingType(reader));
                    }
                }
            }
            catch (SqlException)
            {
                return null;
            }
            catch (Exception)
            {
                return null;
            }
            finally
            {
                bd.CloseConection();
            }
            return subTasks;
        }
        public override SubTask GetById(int id)
        {
            SubTask subTask = null;
            try
            {
                if (id <= 0) return null;
                bd.OpenConection();
                SqlCommand cmd = new SqlCommand("SELECT * FROM [dbo].[SubTask] WHERE Id_SubTask= @Id_SubTask", bd.connection);
                cmd.Parameters.AddWithValue("@Id_SubTask", id);
                using (var reader = cmd.ExecuteReader())
                {
                    if (reader.Read())
                    {
                        subTask = MappingType(reader);
                    }
                }
            }
            catch (SqlException)
            {
                return null;
            }
            catch (Exception)
            {
                return null;
            }
            finally
            {
                bd.CloseConection();
            }
            return subTask;
        }
        public List<SubTask> GetAllByTaskId(int taskId)
        {
            List<SubTask> subtasks = new List<SubTask>();
            try
            {
                bd.OpenConection();
                SqlCommand cmd = new SqlCommand("SELECT * FROM [dbo].[SubTask] WHERE Id_Task = @Id_Task", bd.connection);
                cmd.Parameters.AddWithValue("@Id_Task", taskId);
                using (var reader = cmd.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        subtasks.Add(MappingType(reader));
                    }
                }
            }
            catch (SqlException)
            {
                return null;
            }
            catch (Exception)
            {
                return null;
            }
            finally
            {
                bd.CloseConection();
            }
            return subtasks;
        }
        public override SubTask MappingType(SqlDataReader reader)
        {
            return new SubTask
            {
                id = reader.GetInt32(reader.GetOrdinal("Id_SubTask")),
                Title = reader.GetString(reader.GetOrdinal("Title")),
                Description = reader.IsDBNull(reader.GetOrdinal("Description")) ? null : reader.GetString(reader.GetOrdinal("Description")),
                CreationDate = reader.GetDateTime(reader.GetOrdinal("CreationDate")),
                EndDate = reader.GetDateTime(reader.GetOrdinal("EndDate")),
                State = reader.GetBoolean(reader.GetOrdinal("State")),
                Task = taskRepository.GetById(reader.GetInt32(reader.GetOrdinal("Id_Task")))
            };
        }

        public override bool Update(SubTask entity)
        {
            try
            {
                bd.OpenConection();
                SqlCommand cmd = new SqlCommand(@"UPDATE [dbo].[SubTask] SET 
                Title=@Title, Description=@Description, CreationDate=@CreationDate, 
                EndDate=@EndDate, State=@State, Id_Task=@Id_Task WHERE Id_SubTask=@Id_SubTask", bd.connection);

                cmd.Parameters.AddWithValue("@Id_SubTask", entity.id);
                cmd.Parameters.AddWithValue("@Title", entity.Title);
                cmd.Parameters.AddWithValue("@Description", string.IsNullOrEmpty(entity.Description) ? DBNull.Value : (object)entity.Description);
                cmd.Parameters.AddWithValue("@CreationDate", entity.CreationDate);
                cmd.Parameters.AddWithValue("@EndDate", entity.EndDate);
                cmd.Parameters.AddWithValue("@State", entity.State);
                cmd.Parameters.AddWithValue("@Id_Task", entity.Task.id);

                return cmd.ExecuteNonQuery() > 0;
            }
            catch (SqlException)
            {
                return false;
            }
            catch (Exception)
            {
                return false;
            }
            finally
            {
                bd.CloseConection();
            }
        }
    }
}

File: repository/SubTaskRepository.js
import BaseRepository from "./BaseRepository.js";

const mapFromDb = (row) => ({
  id_SubTask: row.id,
  title: row.title,
  description: row.description,
  creationDate: row.created_at,
  endDate: row.due_date,
  id_Priority: row.priority_id,
  id_Category: row.category_id,
  state: row.completed,
  id_Task: row.parent_task_id,
  id_User: row.user_id,
});

const mapToDb = (entity) => ({
  title: entity.title,
  description: entity.description ?? null,
  due_date: entity.endDate ?? null,
  priority_id: entity.id_Priority ?? null,
  category_id: entity.id_Category ?? null,
  completed: entity.state ?? false,
  user_id: entity.id_User,
  parent_task_id: entity.id_Task,
});

class SubTaskRepository extends BaseRepository {
  constructor() {
    super("tasks", {
      primaryKey: "id",
      mapFromDb,
      mapToDb,
    });
  }

  async save(subTask) {
    return super.save(subTask);
  }

  async getAllByTaskId(taskId) {
    if (!taskId) return [];
    const { data, error } = await this.client
      .from(this.tableName)
      .select("*")
      .eq("parent_task_id", taskId)
      .order("created_at", { ascending: true });

    if (error) {
      throw new Error(error.message);
    }

    return data.map(mapFromDb);
  }

  async getById(id) {
    return super.getById(id);
  }

  async update(subTask) {
    if (!subTask?.id_SubTask) return null;
    return super.update(subTask.id_SubTask, subTask);
  }

  async delete(id) {
    if (!id) return false;
    const { error } = await this.client.from(this.tableName).delete().eq("id", id);
    if (error) {
      throw new Error(error.message);
    }
    return true;
  }

  async markAllAsCompleted(taskId) {
    if (!taskId) return false;
    const { error } = await this.client
      .from(this.tableName)
      .update({ completed: true })
      .eq("parent_task_id", taskId);
    if (error) {
      throw new Error(error.message);
    }
    return true;
  }
}

export default SubTaskRepository;
